testcode/lib/mouse-test.c

```diff
@@ -0,0 +1,431 @@
+        ** message before they remove the current screen.
+        */
+        if (doesclrscrafterexit ()) {
+            cgetc ();
+        }
+        exit (EXIT_FAILURE);
+    }
+}
+
+
+
+#if DYN_DRV
+
+/* Points to the dynamic driver's name. */
+static const char *mouse_name;
+
+
+static void DoWarning (void)
+/* Warn the user that a driver is needed for this program. */
+{
+    cprintf ("Warning: This program needs\r\n"
+             "the driver with the name\r\n"
+             "    %s\r\n"
+             "on a disk! Press 'y' if you have it;\r\n"
+             "or, any other key to exit.\r\n", mouse_stddrv);
+    if (tolower (cgetc ()) != 'y') {
+        exit (EXIT_SUCCESS);
+    }
+    cprintf ("OK. Please wait patiently...\r\n");
+}
+
+#else
+
+unsigned char *mouse_drv_use;
+#endif
+
+
+#ifdef __ATARI__
+#ifdef __ATARIXL__
+#define MSENAME_EXT "X"
+#define MSESTAT_0 atrxjoy_mou
+#define MSESTAT_1 atrxst_mou
+#define MSESTAT_2 atrxami_mou
+#define MSESTAT_3 atrxtrk_mou
+#define MSESTAT_4 atrxtt_mou
+#else
+#define MSENAME_EXT ""
+#define MSESTAT_0 atrjoy_mou
+#define MSESTAT_1 atrst_mou
+#define MSESTAT_2 atrami_mou
+#define MSESTAT_3 atrtrk_mou
+#define MSESTAT_4 atrtt_mou
+#endif
+#define MSENAME_0 "ATR" MSENAME_EXT "JOY.MOU"
+#define MSENAME_1 "ATR" MSENAME_EXT "ST.MOU"
+#define MSENAME_2 "ATR" MSENAME_EXT "AMI.MOU"
+#define MSENAME_3 "ATR" MSENAME_EXT "TRK.MOU"
+#define MSENAME_4 "ATR" MSENAME_EXT "TT.MOU"
+#elif defined(__C64__) || defined(__C128__)
+#ifdef __C64__
+#define MSENAME_EXT "c64-"
+#define MSESTAT_0 c64_joy_mou
+#define MSESTAT_1 c64_1351_mou
+#define MSESTAT_2 c64_inkwell_mou
+#define MSESTAT_3 c64_pot_mou
+#else
+#define MSENAME_EXT "c128-"
+#define MSESTAT_0 c128_joy_mou
+#define MSESTAT_1 c128_1351_mou
+#define MSESTAT_2 c128_inkwell_mou
+#define MSESTAT_3 c128_pot_mou
+#endif
+#define MSENAME_0 MSENAME_EXT "joy.mou"
+#define MSENAME_1 MSENAME_EXT "1351.mou"
+#define MSENAME_2 MSENAME_EXT "inkwell.mou"
+#define MSENAME_3 MSENAME_EXT "pot.mou"
+#endif
+
+
+static void __fastcall__ ShowState (unsigned char Jailed, unsigned char Invisible)
+/* Display jail and cursor states. */
+{
+    cclearxy (0, 7, 32);
+    gotoxy (0, 7);
+    cprintf ("Pointer is %svisible%s.", Invisible? "in" : "", Jailed? " and jailed" : "");
+}
+
+#ifdef __ATARIXL__
+extern char _HIDDEN_RAM_SIZE__, _HIDDEN_RAM_LAST__, _HIDDEN_RAM_START__;
+#endif
+
+#if DYN_DRV
+int main (int argc, char *argv[])
+#else
+int main (void)
+#endif
+{
+    struct mouse_info info;
+    struct mouse_box full_box, small_box;
+    unsigned char width, height;
+    char C;
+    bool Invisible = true, Done = false, Jailed = false;
+
+#ifdef __ATARIXL__
+    cprintf ("adding heap: $%04X bytes at $%04X\r\n",
+             &_HIDDEN_RAM_SIZE__ - (&_HIDDEN_RAM_LAST__ - &_HIDDEN_RAM_START__),
+             &_HIDDEN_RAM_LAST__);
+
+    _heapadd (&_HIDDEN_RAM_LAST__, (size_t)(&_HIDDEN_RAM_SIZE__ - (&_HIDDEN_RAM_LAST__ - &_HIDDEN_RAM_START__)));
+    cgetc ();
+#endif
```

Comment:
Probably not. Don't know if there are any allocations under the hood during the rest of the program flow.